### PR TITLE
Atx tests

### PR DESCRIPTION
**Pull request by antonlerner**
Added tests to atx flow

### CODE REVIEW

File: activation/activation.go [hunk 1]
```diff
@@ -19,41 +21,67 @@ type MeshProvider interface {
...
-		return err
+	ech := b.epochProvider.Epoch(l)
+	var posAtx *mesh.AtxId = nil
+	if ech > 0 {
+		posAtx, err = b.GetPositioningAtxId(ech - 1)
+		if err != nil {
+			return err
+		}
+	}
+	if prevAtx == nil {
```
Comment (y0sher) [inline activation/activation.go:79]: didn't check it in line 63 ?

File: activation/activation.go [hunk 2]
```diff
@@ -65,35 +93,29 @@ func (b Builder) BuildActivationTx(nipst *nipst.NIPST) error {
...
-	}
 	atxId := atxs[rand.Int31n(int32(len(atxs)))]
 
 	return &atxId, nil
 }
 
 func (b *Builder) GetLastSequence(node mesh.NodeId) uint64 {
 	atxId, err := b.GetPrevAtxId(node)
-	if err != nil {
+	if *atxId == mesh.EmptyAtx {
```
Comment (y0sher) [inline activation/activation.go]: why not compare to null pointer ?

File: activation/activation.go [hunk 3]
```diff
@@ -65,35 +93,29 @@ func (b Builder) BuildActivationTx(nipst *nipst.NIPST) error {
...
 func (b *Builder) GetPrevAtxId(node mesh.NodeId) (*mesh.AtxId, error) {
 	ids, err := b.db.GetNodeAtxIds(node)
-	if err != nil {
-		return nil, err
-	}
-	if len(ids) == 0 {
-		return nil, nil
+
+	if len(ids) == 0 || err != nil {
+		return &mesh.EmptyAtx, err
```
Comment (y0sher) [inline activation/activation.go]: why not return nil ?

File: activation/activation_test.go [hunk 4]
```diff
@@ -20,18 +28,58 @@ func (MeshProviderrMock) GetLatestView() []mesh.BlockID {
...
+	bt []byte
+}
 
-func (NetMock) Broadcast(id string, d []byte) error {
+func (n *NetMock) Broadcast(id string, d []byte) error {
+	n.bt = d
 	return nil
 }
 
+func createDB() {
```
Comment (y0sher) [inline activation/activation_test.go]: remove

File: activation/activation.go [hunk 5]
```diff
@@ -19,41 +21,67 @@ type MeshProvider interface {
...
-func NewBuilder(nodeId mesh.NodeId, db database.DB, net api.NetworkAPI, activeSet ActiveSetProvider, view MeshProvider) *Builder {
+func NewBuilder(nodeId mesh.NodeId, db database.DB, net api.NetworkAPI, activeSet ActiveSetProvider, view MeshProvider, epochDuration EpochProvider) *Builder {
 	return &Builder{
-		nodeId, &ActivationDb{db}, net, activeSet, view,
+		nodeId, &ActivationDb{db}, net, activeSet, view, epochDuration,
 	}
 }
 
-func (b Builder) BuildActivationTx(nipst *nipst.NIPST) error {
+func (b *Builder) BuildActivationTx(nipst *nipst.NIPST) error {
```
Comment (y0sher) [inline activation/activation.go]: it's kinda weird that BuildActivationTx broadcasts the tx. I would expect the name to disclose that.

File: activation/activation.go (hunk 2 above)
Comment (y0sher) [reply activation/activation.go]: I'd also check the error beforehand and assume atxId is nil/empty if err is not nil.

Comment (y0sher) [reply activation/activation.go]: also does error always imply the seq number is 0 ? or maybe there's a real error that we need to take care of ?

File: mesh/activation.go [hunk 6]
```diff
@@ -29,6 +29,8 @@ type AtxId struct {
 	common.Hash
 }
 
+var EmptyAtx = AtxId{common.Hash{0}}
```
Comment (y0sher) [inline mesh/activation.go:32]: see above comments, is this necessary?

File: activation/activation.go (hunk 1 above)
Comment (antonlerner) [reply activation/activation.go:79]: please see new code flow

**Review by y0sher — APPROVED** [review]

**Pull request merged**